NIFI-4188 - Nifi RethinkDB Get processor

mans2singh opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

jvwing commented:
Reviewing...

jvwing commented on nifi-nar-bundles/nifi-rethinkdb-bundle/nifi-rethinkdb-processors/src/main/java/org/apache/nifi/processors/rethinkdb/GetRethinkDB.java:
```diff
@@ -0,0 +1,197 @@
+            Map<String,Object> document = getDocument(id, readMode);
+
+            if ( document == null ) {
+                getLogger().error("Document not found for id " + id);
```
Is not found an error?  It seems possible that this could happen at high volume.  Given that it has it's own relationship, and there is a separate relationship for failure, didn't this "succeed" in not finding the key?  I don't have a strong opinion, I'll accept it as-is if you believe in it.

mans2singh commented:
Hello @jvwing - I've updated the code based on your review comments (updated notice, refactored constants to base class).  

Regarding the name Get vs Fetch - I am trying to use the terminology used by RethinkDB and their apis.  If you think that Fetch is more expressive and I can change it. Also, I've added tag 'fetch' so that the processor will turn up in searches.

I was interpolating id in case of empty id is because sometimes it might be null.  So, I've updated the code to check blank rather than just empty and have also added unit tests for those scenarios.

Regarding 'not_found' relation - I wanted to give the users option to handle this scenario separately than other errors.  From what I understand, the user may connect both 'failure' and 'not_found' to same component without losing any flexibility.

Please let me know if you have any other recommendations/thoughts.

Thanks again.

Mans

jvwing commented on nifi-nar-bundles/nifi-rethinkdb-bundle/nifi-rethinkdb-processors/src/main/java/org/apache/nifi/processors/rethinkdb/GetRethinkDB.java:
```diff
@@ -137,9 +135,9 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-        if ( StringUtils.isEmpty(id) ) {
-            getLogger().error("Empty id '" + id + "'");
-            flowFile = session.putAttribute(flowFile, RETHINKDB_ERROR_MESSAGE, "Empty id '" + id + "'");
+        if ( StringUtils.isBlank(id) ) {
```
I do not recommend this change to using `isBlank()`, because it will return true if the id is whitespace only, and I believe whitespace is in fact a valid RethinkDB key.

I'm sorry I was not clear about my concern.  In the earlier code using `isEmpty()`, the if block would be entered if id was null or an empty string.  However, property values are not null, they would be an empty string at least.  So the block appeared to only be entered in practice if id was an empty string.  Given that we knew id was always an empty string, it did not seem necessary to interpolate the value.  It is an empty string, and the result is always the same.

This appeared to be confirmed by your testBlankId() unit test that exercised this code and checked its value against `Blank id ''`.  It appears to be further confirmed by the new testNullId() unit test that checks against the exact same value. 

How about just logging it as `Document Identifier cannot be empty` or similar?

jvwing commented:
My question about the not_found relationship is not if we should have it or not -- I think it is a great addition.  My question is about the logging for that code path.  You log an error on line 150 when a flowfile is routed to not_found, and I'm asking if you believe not_found is "bad enough" to deserve an error.  It seems possible that it is a perfectly normal occurrence that keys are not found in the table, and the flowfile would be routed to not_found without any worries.  As you point out, the user can choose to treat not_found separately from failure, or not, as they choose.  But it is logged as an error regardless of their choice.

jvwing commented:
I'm OK with the name "GetRethinkDB".  Thanks for the tagging.

mans2singh commented:
Hi @jvwing - 

Thanks for your review.

I've updated the code based on your review (changed logging level for not_found to debug and id check to empty + updates to error message and tests).  

Please let me know if there is any other recommendation.

Mans

jvwing commented:
Thanks for those updates, @mans2singh, everything looks good.  I'll merge shortly.

mans2singh commented:
Thanks @jvwing for your advice and guidance.

Mans
